builtin/provisioners/chef/resource_provisioner.go

```diff
@@ -334,44 +359,79 @@ func applyFn(ctx context.Context) error {
 	}
 	defer once.Do(cleanupUserKey)
 
-	if !p.SkipInstall {
-		if err := p.installChefClient(o, comm); err != nil {
+	if attempt == 0 {
+		if !p.SkipInstall {
+			if err := p.installChefClient(o, comm); err != nil {
+				return err
+			}
+		}
+
+		o.Output("Creating configuration files...")
+		if err := p.createConfigFiles(o, comm); err != nil {
 			return err
 		}
-	}
 
-	o.Output("Creating configuration files...")
-	if err := p.createConfigFiles(o, comm); err != nil {
-		return err
-	}
+		if !p.SkipRegister {
+			if p.FetchChefCertificates {
+				o.Output("Fetch Chef certificates...")
+				if err := p.fetchChefCertificates(o, comm); err != nil {
+					return err
+				}
+			}
 
-	if !p.SkipRegister {
-		if p.FetchChefCertificates {
-			o.Output("Fetch Chef certificates...")
-			if err := p.fetchChefCertificates(o, comm); err != nil {
+			o.Output("Generate the private key...")
+			if err := p.generateClientKey(o, comm); err != nil {
 				return err
 			}
 		}
 
-		o.Output("Generate the private key...")
-		if err := p.generateClientKey(o, comm); err != nil {
-			return err
+		if p.Vaults != nil {
+			o.Output("Configure Chef vaults...")
+			if err := p.configureVaults(o, comm); err != nil {
+				return err
+			}
 		}
+
+		// Cleanup the user key before we run Chef-Client to prevent issues
+		// with rights caused by changing settings during the run.
+		once.Do(cleanupUserKey)
+
+		o.Output("Starting initial Chef-Client run...")
 	}
 
-	if p.Vaults != nil {
-		o.Output("Configure Chef vaults...")
-		if err := p.configureVaults(o, comm); err != nil {
+	if err := p.runChefClient(o, comm); err != nil {
+		// Allow RFC062 Exit Codes
+		// https://github.com/chef/chef-rfc/blob/master/rfc062-exit-status.md
+		exitError, ok := err.(*remote.ExitError)
+		if !ok {
+			return fmt.Errorf("Expected remote.ExitError but got: %w", err)
+		}
+		exitStatus := exitError.ExitStatus
+		switch exitStatus {
+		case 35:
+			o.Output("Reboot has been scheduled in the run state")
+			err = nil
+		case 37:
+			o.Output("Reboot needs to be completed")
+			err = nil
+		case 213:
+			o.Output("Chef has exited during a client upgrade")
+			err = nil
+		}
+
+		if len(p.RetryOnExitCode) == 0 || attempt == p.MaxRetries {
 			return err
 		}
-	}
 
-	// Cleanup the user key before we run Chef-Client to prevent issues
-	// with rights caused by changing settings during the run.
-	once.Do(cleanupUserKey)
+		for _, code := range p.RetryOnExitCode {
+			if code == exitStatus {
+				o.Output(fmt.Sprintf("Waiting %v seconds before reconnecting & re-running Chef...", p.WaitForRetry))
+				time.Sleep(time.Duration(p.WaitForRetry) * time.Second)
+				attempt++
+				goto retry
```

Comment:
Interesting way to solve this... But to be honest not one that I like or which seems idiomatic. If you are OK with it, I can make a suggestion to do this is a different way. Using `goto` is not something I would like to use and merge for this.

Reply:
I'm ok with whatever method is more idiomatic @svanharmelen

Reply:
I couldn't push to your branch, so instead I made a PR: https://github.com/bdwyertech/terraform/pull/1 Please have a look and if your OK with it merge it so it will become part of this PR. Thanks!

Reply:
Merged -- Thanks @svanharmelen

Reply:
Cool. I asked a colleague if he could do one last check, and I guess we should be good to go after that...

Reply:
@shoekstra found a few small issues when testing, so we created a second PR (https://github.com/bdwyertech/terraform/pull/2) to your branch with some final tweaks. We tested a scenarios we could think of, so if you merge the PR I will merge this PR over here...

Reply:
Your too fast! I noticed one more thing in the docs... Will be PR #3 I guess. Hold on...